src/Cardano/Wallet/Mnemonic.hs

```diff
@@ -0,0 +1,332 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- | Module provides mnemonic creation and
+-- | restoring from backup phrase functionality
+
+module Cardano.Wallet.Mnemonic
+       (
+       -- * Types
+         Mnemonic
+       , Entropy
+       , EntropySize
+       , MnemonicWords
+
+       -- * Errors
+       , MnemonicError(..)
+       , MnemonicException(..)
+       , EntropyError(..)
+       , DictionaryError(..)
+       , MnemonicWordsError(..)
+
+       -- * Creating @Mnemonic@ (resp. @Entropy@)
+       , mkEntropy
+       , mkMnemonic
+       , genEntropy
+
+       -- * Converting from and to @Mnemonic@ (resp. @Entropy@)
+       , mnemonicToEntropy
+       , entropyToMnemonic
+       , entropyToByteString
+       ) where
+
+
+import Prelude
+
+
+import Basement.Sized.List
+    ( unListN )
+import Control.Arrow
+    ( left )
+import Control.Monad
+    ( when, (>=>) )
+import Control.Monad.Catch
+    ( throwM )
+import Crypto.Encoding.BIP39
+import Data.Aeson
+    ( FromJSON (..), ToJSON (..) )
+import Data.Aeson.Types
+    ( Parser )
+import Data.ByteArray
+    ( constEq )
+import Data.ByteString
+    ( ByteString )
+import Data.Default
+    ( Default (def) )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Text
+    ( Text, pack, unpack )
+import Data.Typeable
+    ( Typeable )
+import Data.Word
+    ( Word8 )
+import GHC.TypeLits
+    ( KnownNat, Nat, natVal )
+import Test.QuickCheck
+    ( Arbitrary (..) )
+import Test.QuickCheck.Gen
+    ( vectorOf )
+
+
+import qualified Basement.Compat.Base as Basement
+import qualified Basement.String as Basement
+import qualified Basement.UArray as Basement
+import qualified Crypto.Encoding.BIP39.English as Dictionary
+import qualified Crypto.Random.Entropy as Crypto
+import qualified Data.ByteString.Char8 as B8
+
+
+
+-- | A backup-phrase in the form of a non-empty of Mnemonic words
+-- Constructor isn't exposed.
+data Mnemonic (mw :: Nat) = Mnemonic
+    { mnemonicToEntropy  :: Entropy (EntropySize mw)
+    , mnemonicToSentence :: MnemonicSentence mw
+    } deriving (Eq, Show)
+
+
+newtype MnemonicException csz =
+    UnexpectedEntropyError (EntropyError csz)
+    deriving (Show, Typeable)
+
+
+data MnemonicError csz
+    = ErrMnemonicWords MnemonicWordsError
+    | ErrEntropy (EntropyError csz)
+    | ErrDictionary DictionaryError
+    | ErrForbidden
+    deriving (Show)
+
+
+
```

Comment:
Lots of extra blank lines here! Perhaps just have one?

Reply:
done